### ISSUE TITLE
Type inference on methods returning $this

### ISSUE THREAD

**Issue by tdtm** [vimeo/psalm#4263]
When using chained method calls returning `$this`, an error is raised for "UndefinedMethod".

Samples:
- With return type `self`; no phpdoc: https://psalm.dev/r/3d454e34c2
	Doesn't work.
- With return type of classname; no phpdoc: https://psalm.dev/r/23a3e1ee01
	Doesn't work.
- With return type `self`; AND `@return $this` phpdoc: https://psalm.dev/r/bda78a0789
	Works.
- With return type `self`; AND `@return static` phpdoc: https://psalm.dev/r/86e41520e7
	Works.

It also won't use `return $this` to infer the return type if no return type is provided.

tl;dr: It works if phpdoc is provided, but it won't use `return $this` to clarify.

**Comment by weirdan** [comment]
Can you elaborate on *why* you think it should work differently?

**Comment by tdtm** [comment]
I'll do my best!

- The code works (https://3v4l.org/Lc73L)
- PhpStorm infers it successfully
- This is somewhat common in libraries using fluent patterns. In my case, this was prompted by issues raised when using specifically https://github.com/spatie/icalendar-generator
- It works with the phpdoc specifying what's essentially the same as the return (`/** @return $this */`), so having to redundantly provide the PhpDoc from what could be inferred seems... inefficient? And it appears that it _can_ be understood.

**Comment by weirdan** [comment]
If Psalm inferred the return type as `$this` here: https://psalm.dev/r/392e337d4d it would have to consider this: https://psalm.dev/r/f12b8057e3 an LSP violation, and there's clearly none. Returning `$this` in `C::ugh()` is an implementation detail. Since this would introduce ambiguity I don't think this is feasible, nor, in fact, desirable.

**Comment by muglug** [comment]
Yeah, Psalm has followed the rule that it infers types only inside functions, and I don't think that's worth violating. PHP 8 supports a `static` return type (which Psalm will support fully) to help with this particular use-case.

**Comment by tdtm** [comment]
I must admit I hadn't considered PHP8's changes, which do solve the issue.

Thanks to both of you for your quick replies, it's much appreciated!

So as solution (until PHP8), in any of our code (not that we have much fluent code tbh) if that arises our best bet would be to use phpdoc with @return static.

If I may end with a quick question: any tip you could suggest that could be used in consumer code (ie when consuming an external package) to solve that? 

- I tried using a .phpstorm.meta.php file to type-hint as 'static', without any effect. (IIRC they're taken into account!?) 
- Adding to baseline to ignore the issue is kinda dirty. 
- The only solution I came up with was to break down the fluent calls in step-by-step.... 

....or PR into that external package to add the static @return annotation 🤷‍♂️

**Comment by andrew-demb** [comment]
@tdtm you can use stub files https://psalm.dev/docs/running_psalm/plugins/authoring_plugins/#stub-files

Or send PR to package :)

**Comment by tdtm** [comment]
Thanks @andrew-demb. I'll check with maintainers if they'll accept a PR for it.
